letsencrypt/account.py

```diff
@@ -22,212 +23,179 @@
-    :type key: :class:`~letsencrypt.le_util.Key`
-
-    :ivar str email: Client's email address
-    :ivar str phone: Client's phone number
+    :ivar .RegistrationResource regr: Registration Resource
+    :ivar .JWK key: Authorized Account Key
+    :ivar .Meta: Account metadata
 
-    :ivar regr: Registration Resource
-    :type regr: :class:`~acme.messages.RegistrationResource`
+    .. note:: ``creation_dt`` and ``creation_host`` are useful in
+        cross-machine migration scenarios.
 
     """
 
-    # Just make sure we don't get pwned
-    # Make sure that it also doesn't start with a period or have two consecutive
-    # periods <- this needs to be done in addition to the regex
-    EMAIL_REGEX = re.compile("[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+$")
+    class Meta(jose.JSONObjectWithFields):
+        """Account metadata
 
-    def __init__(self, config, key, email=None, phone=None, regr=None):
-        le_util.make_or_verify_dir(
-            config.accounts_dir, 0o700, os.geteuid())
-        self.key = key
-        self.config = config
-        if email is not None and self.safe_email(email):
-            self.email = email
-        else:
-            self.email = None
-        self.phone = phone
+        :ivar datetime.datetime creation_dt: Creation date and time (UTC).
+        :ivar str creation_host: FQDN of host, where account has been created.
+
+        """
+        creation_dt = acme_fields.RFC3339Field("creation_dt")
+        creation_host = jose.Field("creation_host")
 
+    def __init__(self, regr, key, meta=None):
+        self.key = key
         self.regr = regr
+        self.meta = self.Meta(
+            # pyrfc3339 drops microseconds, make sure __eq__ is sane
+            creation_dt=datetime.datetime.now(
+                tz=pytz.UTC).replace(microsecond=0),
+            creation_host=socket.getfqdn()) if meta is None else meta
 
     @property
-    def uri(self):
-        """URI link for new registrations."""
-        if self.regr is not None:
-            return self.regr.uri
-        else:
-            return None
+    def id(self):  # pylint: disable=invalid-name
+        """Globally unique account identifier.
 
-    @property
-    def new_authzr_uri(self):  # pylint: disable=missing-docstring
-        if self.regr is not None:
-            return self.regr.new_authzr_uri
-        else:
-            return None
+        Implementation note: Email? Multiple accounts can have the same
+        email address. Registration URI? Assigned by the server, not
+        guaranteed to be stable over time, nor cannonical URI can be
+        generated. One could use the account key (fingerprint), as ACME
+        protocol doesn't allow it to be updated...
 
-    @property
-    def terms_of_service(self):  # pylint: disable=missing-docstring
-        if self.regr is not None:
-            return self.regr.terms_of_service
-        else:
-            return None
+        """
+        return self.slug
```

Comment:
Sorry missed that it was a property.  I do think that the hash of the account key is a bit easier to understand.